src/v/cloud_storage_clients/client_pool.cc

```diff
@@ -54,11 +62,58 @@ client_pool::acquire(ss::abort_source& as) {
             co_await wait_for_credentials();
         }
 
-        while (_pool.empty() && !_gate.is_closed() && !_as.abort_requested()) {
+        while (unlikely(
+          _pool.empty() && !_gate.is_closed() && !_as.abort_requested())) {
             if (_policy == client_pool_overdraft_policy::wait_if_empty) {
                 co_await _cvar.wait();
+                vlog(
+                  pool_log.debug,
+                  "cvar triggered, pool size: {}",
+                  _pool.size());
             } else {
-                _pool.emplace_back(make_client());
+                // Ask other shards for allowance
+                auto resources = co_await container().map(
```

Comment:
Yes and no. When the `_pool` is empty it will broadcast to find the candidate to borrow from. Then it will try to borrow and if it fail it will wait on a `_cvar` old style until the client will be returned into the `_pool`. There is very high probability that the connection will be acquired after that because `_cvar` notifies waiters one by one using `signal`. It doesn't notify all of them by doing `broadcast`. So it is possible for the system to exhibit the behaviour you described but it's unlikely. 

Also, I'm thinking that maybe doing `invoke_on_all` is a bit expensive and instead I can do something a bit more smart. For instance, 2-random strategy could be used when 2 random shards to borrow from are chosen and the one with least utilisation is selected.